allennlp/training/trainer.py

```diff
@@ -973,11 +973,7 @@ def from_partial_objects(
         if not optimizer_:
             optimizer_ = Optimizer.default(parameters)
 
-        try:
-            batches_per_epoch = len(data_loader)
-        except TypeError:
-            # If the dataset is lazy, it won't have a length.
-            batches_per_epoch = None
+        batches_per_epoch = len(data_loader)  # returns "1" instead of TypeError for _LazyInstances
```

Comment:
The ```TypeError``` exception here is never actually thrown since the ``__len__()`` method is overridden in AllenNLP's ```_LazyInstances``` and always returns ```1```. 
In case someone somehow passes an instance of plain ```IterableDataset```, it is better to actually let it throw TypeError then pass a custom ```None``` to the scheduler which it cannot handle it anyway (it accepts ```int```).

Reply:
I agree with your analysis of the behavior here, but does having a length of 1 make sense for the learning rate schedulers?

Reply:
If you ask me, having the length on ```1``` does not make sense in all cases apart from the case where we have one batch per epoch. 

Here is our doc for the method:
https://github.com/allenai/allennlp/blob/d66db44dfe8072f0b8bac2770968d996664e683b/allennlp/data/dataset_readers/dataset_reader.py#L93-L100
I would suggest to actually do crash instead and make clients of the method handle the exception in the proper way.
  
---
Regarding Schedulers, the only one which asks for this parameter is currently is ```SlantedTriangular```. I didn't look closely, but it seems it has some proper response to cases like this:
https://github.com/allenai/allennlp/blob/d66db44dfe8072f0b8bac2770968d996664e683b/allennlp/training/learning_rate_schedulers/slanted_triangular.py#L139
but it might be by chance. 
---

Reply:
That like is not "by chance", but it is a bit of defensive coding that should not be relied on. I believe if `actual_num_steps_per_epoch` ends up smaller than the real number of steps, your learning rate would go negative. I wanted to avoid that with this line.

Reply:
Thanks for the note.
> That line is not "by chance", but it is a bit of defensive coding

Turns out we discovered that unfortunately, IterableDatasets with undefined length breach the defense :)

Anyway, it existed before and this PR does not change the interaction between _LazyInstances and the scheduler, so we could have a separate issue.